src/bin/pgcopydb/table-data.c

```diff
@@ -836,6 +840,57 @@ copydb_table_data_worker(CopyDataSpec *specs)
 }
 
 
+typedef struct CopyProgressContext
+{
+	CopyDataSpec *specs;
+	CopyTableDataSpec *tableSpecs;
+	uint64_t bytesTransmitted;
+	uint64_t bytesTransmittedBeforeSavingProgress; /* the bytes transmitted before saving the progress to DB */
+	instr_time lastSavingTimeMs; /* the last saving time of the progress to the DB */
+} CopyProgressContext;
+
+/*
+ * on_copy_progress_hook saves the copy progress into the catalog.
+ * It writes to the DB within at an arbitrary interval.
+ */
+static bool
+on_copy_progress_hook(int bytesTransmitted, void *context)
+{
+	CopyProgressContext *ctx = (CopyProgressContext *) context;
+	CopyDataSpec *specs = ctx->specs;
+	CopyTableDataSpec *tableSpecs = ctx->tableSpecs;
+
+	ctx->bytesTransmittedBeforeSavingProgress += bytesTransmitted;
+	ctx->bytesTransmitted += bytesTransmitted;
```

Comment:
When are the two values different? There is no comment about that, and the naming of the variable does not explain that either. Why do we need the two slots? Can we have a better name for `bytesTransmittedBeforeSavingProgress`? I currently have two problems with it: I don't understand the name in relation with the code, and I would like a shorted one if that's possible.

Reply:
Just because you don't understand it doesn't mean others don't, right? So the variable keeps the number of bytes transmitted before saving the progress to the db and reset to zero on each save. If you have a better name, please feel free to suggest one.
Assuming that you mean the vars starting with `bytesTransmitted...` one keeps the total as before (no change here) and the other as explained above.

Reply:
Is that attitude necessary @rimbi ? I can tell you it's not welcome.

Two things for you to consider:
  1. I am asking questions on aspects of your code that is not obvious to me when reviewing the PR. When things are not obvious, we need to either improve the code, or add good comments to help the reader understand what's happening with less efforts.
  2. I have read a lot of code, including some great code in many different programming languages. The great code that I have read always have these two properties: what/how it does is obvious, and it is elegant. Elegance has something personal to it, so I will merge code that I do not find elegant. I will not merge code that I do not find obvious, though. That's not personal.